csrc/ops/arith.cpp

```diff
@@ -2684,4 +2684,54 @@ TensorView* tensor(Val* val) {
   return out;
 }
 
+namespace {
+
+//! Create new output for matmul
+static TensorView* newForMatmul(
+    TensorView* tv_a,
+    TensorView* tv_b
+) {
+  auto orig_domain_a = TensorDomain::noReductions(tv_a->getMaybeRFactorDomain());
+  auto orig_domain_b = TensorDomain::noReductions(tv_b->getMaybeRFactorDomain());
+  auto ndims_a = orig_domain_a.size();
+  auto ndims_b = orig_domain_b.size();
+  auto ndims_out = std::max(ndims_a, ndims_b);
+  NVF_ERROR(ndims_a>=1 && ndims_b>= 1);
+
+  std::vector<IterDomain*> out_domain(ndims_out, nullptr);
+  
+  int kpos_a = ndims_a - 1;
+  int kpos_b = ndims_b > 1 ? ndims_b - 2 : 0;
+
+  for (int inx = ndims_out - 1, inx_a = ndims_a - 1, inx_b = ndims_b - 1; inx >= 0; inx--, inx_a--, inx_b--){
+    std::vector<IterDomain*> input_ids;
+    if (inx_a >= 0 && inx_a != kpos_a){
+      input_ids.emplace_back(orig_domain_a[inx_a]);
+    }
+    if (inx_b >= 0 && inx_b != kpos_b){
+      input_ids.emplace_back(orig_domain_b[inx_b]);
+    }
+    out_domain[inx] = ops::outIterDomain(input_ids);
+  }  
+
+  TensorDomain* td = IrBuilder::create<TensorDomain>(
+      out_domain, TensorDomain::getContiguityFilledWith(out_domain, true));
+
+  return IrBuilder::create<TensorView>(td, tv_a->dtype());
+}
+
+} // namespace
+
+
+// TODO (Priya): This will be renamed to matmul once we are ready to modify the python API backend.
+// Keeping separate for now, to avoid breaking tests in Thunder.
+TensorView* eagerMatmul(
+    TensorView* tv_a,
+    TensorView* tv_b) {
+  NVF_CHECK(tv_a->getDataType().value() == tv_b->getDataType().value());
+  TensorView* out = newForMatmul(tv_a, tv_b);
```

Comment:
In a followup PR I think we should avoid creating this node for the following two special cases:
1. Both inputs are 1D. We should `return sum(mul(tv_a, tv_b));`.
2. Both inputs are `IterType::Broadcast` in the K dimension i.e. `tv_a [M, 1]` and `tv_b [1, N]`. In this case the output is a simple outer product `return mul(tv_a, tv_b);`.

Avoiding `MatmulOp`s means it will be easier for us to fuse into the reduction or pointwise schedulers.

Reply:
Added the first case to this PR.

Reply:
In these cases, do we add a `castOp` as well? `castOp` is needed to mimic `torch.matmul` behavior.

Reply:
No, I think we should just create the output as half-precision instead of `Float`.

Reply:
Oh I see what you mean. Yeah I think at first we'd need to insert the cast.

Reply:
Added the first case to this PR.